modules/platforms/python/pyignite/api/key_value.py

```diff
@@ -0,0 +1,1105 @@
+from pyignite.queries import Query, Response
+from .result import APIResult
+from pyignite.utils import cache_id
+
+
+def cache_put(
+    connection: 'Connection', cache: Union[str, int], key, value,
+    key_hint=None, value_hint=None, binary=False, query_id=None,
+) -> APIResult:
+    """
+    Puts a value with a given key to cache (overwriting existing value if any).
+
+    :param connection: connection to Ignite server,
+    :param cache: name or ID of the cache,
+    :param key: key for the cache entry. Can be of any supported type,
+    :param value: value for the key,
+    :param key_hint: (optional) Ignite data type, for which the given key
+     should be converted,
+    :param value_hint: (optional) Ignite data type, for which the given value
+     should be converted.
+    :param binary: (optional) pass True to keep the value in binary form.
+     False by default,
+    :param query_id: (optional) a value generated by client and returned as-is
+     in response.query_id. When the parameter is omitted, a random value
+     is generated,
+    :return: API result data object. Contains zero status if a value
+     is written, non-zero status and an error description otherwise.
+    """
+
+    class CachePutQuery(Query):
+        op_code = OP_CACHE_PUT
+
+    query_struct = CachePutQuery([
+        ('hash_code', Int),
+        ('flag', Byte),
+        ('key', key_hint or AnyDataObject),
+        ('value', value_hint or AnyDataObject),
+    ], query_id=query_id)
+
+    _, send_buffer = query_struct.from_python({
+        'hash_code': cache_id(cache),
+        'flag': 1 if binary else 0,
+        'key': key,
+        'value': value,
+    })
+
+    connection.send(send_buffer)
+
+    response_struct = Response([])
+    response_class, recv_buffer = response_struct.parse(connection)
+    response = response_class.from_buffer_copy(recv_buffer)
+
+    result = APIResult(response)
+    if result.status != 0:
+        return result
+    result.value = response_struct.to_python(response)
+    return result
+
+
+def cache_get(
+    connection: 'Connection', cache: Union[str, int], key,
+    key_hint=None, binary=False, query_id=None,
+) -> APIResult:
+    """
+    Retrieves a value from cache by key.
+
+    :param connection: connection to Ignite server,
+    :param cache: name or ID of the cache,
+    :param key: key for the cache entry. Can be of any supported type,
+    :param key_hint: (optional) Ignite data type, for which the given key
+     should be converted,
+    :param binary: (optional) pass True to keep the value in binary form.
+     False by default,
+    :param query_id: (optional) a value generated by client and returned as-is
+     in response.query_id. When the parameter is omitted, a random value
+     is generated,
+    :return: API result data object. Contains zero status and a value
+     retrieved on success, non-zero status and an error description on failure.
+    """
+
+    class CacheGetQuery(Query):
```

Comment:
I must agree with you on performance, although I believe the impact should be negligible.

Another question is: where to define these classes? Module level is no go: they will pollute the namespace, worsen readability, clutter docs, and make debugging more difficult…

Now I think about parameterizing `Query` with opcode rather than subclassing it.

Reply:
I've write simple test to check performance impact if inner class used:

```
import time

with_class_counter = 0
no_class_counter = 0


class OtherClass:
    a = 1


def with_class():
    global with_class_counter

    class SomeClass:
        a = 1

    a = SomeClass()
    with_class_counter += a.a


def no_class():
    global no_class_counter

    a = OtherClass()
    no_class_counter += a.a


before = int(round(time.time() * 1000))

for i in range(1, 1000000):
    no_class()

after = int(round(time.time() * 1000))

print("Time for method w/o inner class: %s" % str(after-before))


before = int(round(time.time() * 1000))

for i in range(1, 1000000):
    with_class()

after = int(round(time.time() * 1000))

print("Time for method with inner class: %s" % str(after-before))
```

And output:
Time for method w/o inner class: 259
Time for method with inner class: 9220

I think that parametrising Query is good option.